File: cascadingFailuresTaskMigration_python/opt/opt.py
import networkx as nx
from typing import List, Dict
from input.task import Task
from input.robot import Robot
from input.group import Group
from input.experiment_result import ExperimentResult
from main.initialize import Initialize
from evaluation.evalution import Evalution
from MPFTM.finder_leader import FinderLeader
from .opt_migration import OptMigration


class Opt:
    def __init__(self, tasks: List[Task], arc_graph: nx.Graph, robots: List[Robot],
                 a: float, b: float):
        self.tasks = tasks
        self.robots = robots
        self.arc_graph = arc_graph
        self.id_to_groups: Dict[int, Group] = {}
        self.id_to_robots: Dict[int, Robot] = {}
        self.shortest_path_dict: Dict = {}
        self.a = a
        self.b = b

    def opt_run(self) -> ExperimentResult:
        """Run optimization algorithm."""
        print("optRun")

        ini = Initialize()
        evalution = Evalution(self.id_to_robots, self.id_to_groups)
        experiment_result = ExperimentResult()

        for robot in self.robots:
            self.id_to_robots[robot.robot_id] = robot

        ini.run(self.tasks, self.robots, self.id_to_groups, self.id_to_robots)
        self._leader_selection(self.id_to_groups, self.id_to_robots, self.arc_graph)

        # Calculate all shortest paths
        self.shortest_path_dict = dict(nx.all_pairs_dijkstra_path_length(self.arc_graph, weight='weight'))

        opt_migration = OptMigration(
            self._convert_shortest_path_dict(), self.id_to_groups,
            self.id_to_robots, self.a, self.b
        )
        migration_records = opt_migration.run()

        sum_migration_cost = evalution.calculate_migration_cost(
            self._convert_shortest_path_dict(), migration_records
        )
        sum_execute_cost = evalution.calculate_execute_tasks_cost(self.robots)
        survival_rate = evalution.calculate_mean_survival_rate(self.robots)

        experiment_result.mean_execute_cost = sum_execute_cost
        experiment_result.mean_survival_rate = survival_rate
        experiment_result.mean_migration_cost = sum_migration_cost

        return experiment_result

    def _convert_shortest_path_dict(self) -> Dict:
        """Convert NetworkX shortest path dict to flat dict with (from, to) keys."""
        result = {}
        for source, targets in self.shortest_path_dict.items():
            for target, length in targets.items():
                result[(source, target)] = length
        return result

    def _leader_selection(self, id_to_groups: Dict[int, Group], id_to_robots: Dict[int, Robot],
                         arc_graph: nx.Graph):
        """Select leader for each group."""
        finder = FinderLeader()
        for group in id_to_groups.values():
            if group.leader is None:
                group.leader = finder.find_leader(group, id_to_robots, id_to_groups,
                                                 arc_graph, self.a, self.b)

        # Add edges between leader nodes
        for group_id in id_to_groups.keys():
            leader_id = id_to_groups[group_id].leader.robot_id
            for to_group_id in id_to_groups.keys():
                to_leader_id = id_to_groups[to_group_id].leader.robot_id
                if group_id != to_group_id and not arc_graph.has_edge(leader_id, to_leader_id):
                    arc_graph.add_edge(leader_id, to_leader_id, weight=10)

File: cascadingFailuresTaskMigration/src/main/java/opt/Opt.java
package opt;

import evaluation.Evalution;
import input.*;
import main.Initialize;
import org.jgrapht.alg.interfaces.ShortestPathAlgorithm;
import org.jgrapht.alg.shortestpath.DijkstraShortestPath;
import org.jgrapht.graph.DefaultUndirectedWeightedGraph;
import org.jgrapht.graph.DefaultWeightedEdge;
import MPFTM.FinderLeader;

import java.util.HashMap;
import java.util.List;

public class Opt {
    //主算法
    private List<Task> tasks;
    private List<Robot> robots;
    private HashMap<Integer, Group> idToGroups;
    private HashMap<Integer,Robot> idToRobots;
    private DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph;
    private ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath ;
    Double a;
    Double b;
    public Opt(List<Task> tasks, DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph,
               List<Robot> robots, Double a, Double b) {
        this.tasks = tasks;
        this.robots = robots;
        this.arcGraph = arcGraph;
        idToGroups = new HashMap<>();
        idToRobots = new HashMap<>();
        this.a = a;
        this.b = b;
    }

    public ExperimentResult optRun() {
        System.out.println("optRun");
        Initialize ini = new Initialize();
        Evalution evalution = new Evalution(idToRobots, idToGroups);
        ExperimentResult experimentResult = new ExperimentResult();
        for (Robot robot : robots) {
            idToRobots.put(robot.getRobotId(), robot);
        }
        ini.run(tasks, robots, idToGroups, idToRobots);
        leaderSelection(idToGroups,idToRobots,arcGraph);

        Double sumExecuteCost;
        Double survivalRate;
        Double sumMigrationCost;

        shortestPath = new DijkstraShortestPath<>(arcGraph);
        List<MigrationRecord> migrationRecords = new OptMigration(shortestPath,idToGroups,idToRobots,a,b).run();
        //搞一个最优分配的方式
        sumMigrationCost = evalution.calculateMigrationCost(shortestPath, migrationRecords);
        sumExecuteCost = evalution.calculateExecuteTasksCost(robots);
        survivalRate = evalution.calculateMeanSurvivalRate(robots);

        experimentResult.setMeanExecuteCost(sumExecuteCost);
        experimentResult.setMeansurvivalRate(survivalRate);
        experimentResult.setMeanMigrationCost(sumMigrationCost);
        return experimentResult;
    }

    private void leaderSelection(HashMap<Integer, Group> idToGroups, HashMap<Integer,Robot> idToRobots, DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph) {
        for (Group group : idToGroups.values()) {
            if (group.getLeader()==null) {
                //对于子图来选取leader节点
                group.setLeader(new FinderLeader().findLeader(group,idToRobots,idToGroups,arcGraph,a,b));
            }
        }
        //给这些leader节点之间添加上连接的边
        for (Integer groupId : idToGroups.keySet()) {
            Integer leaderId = idToGroups.get(groupId).getLeader().getRobotId();
            for (Integer toGroupId : idToGroups.keySet()) {
                Integer toLeaderId = idToGroups.get(toGroupId).getLeader().getRobotId();
                if (!groupId.equals(toGroupId)&&!arcGraph.containsEdge(leaderId,toLeaderId)) {
                    arcGraph.addEdge(leaderId,toLeaderId);
                    arcGraph.setEdgeWeight(leaderId,toLeaderId,10);
                }
            }
        }
    }
}
